src/org/thoughtcrime/securesms/connect/DcEventCenter.java

```diff
@@ -42,6 +44,16 @@ public void addObserver(int eventId, @NonNull DcEventDelegate observer) {
         }
     }
 
+    public void addMultiAccountObserver(int eventId, @NonNull DcEventDelegate observer) {
+        synchronized (MULTI_LOCK) {
+            ArrayList<DcEventDelegate> idObservers = multiObservers.get(eventId);
+            if (idObservers == null) {
+                multiObservers.put(eventId, (idObservers = new ArrayList<>()));
+            }
+            idObservers.add(observer);
+        }
+    }
+
     public void removeObserver(int eventId, DcEventDelegate observer) {
```

Comment:
it is error prone to track exactly if an observer is a multiaccount observer or not and call the correct remover. it would be better to let `removeObserver()` remove both types. otherwise, i already see bugs arising when one changes from single-observer to multi-observer for reasons - but does not change the remover-function.

this is esp. bad as it is not directly visible - but will result in slowdown over time (we had these hard-to-get-errors on iOS at some point)


same for `removeObservers()`, which is a convenience tool to remove just all observers, therefore, it should remove the multiAccountObservers as well.

two function less - and the overhead of one lookup in a most times empty hash is negligible

Reply:
indeed, in practice you rather always stop listening to all events at once